### PR TITLE
get http

### PR DESCRIPTION
**Pull request by LjubanDragosavljevic**

### CODE REVIEW

File: src/app/books/books.service.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { Observable } from 'rxjs';
+import { map, delay } from 'rxjs/operators';
+import { Book } from './book/book.model';
+
+const API_URL =
```
Comment (stpetra) [inline src/app/books/books.service.ts]: type this const (also tip: you can create separate config file where you will declare all constants you will use in you project)

File: src/app/books/books.service.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  'https://api.nytimes.com/svc/books/v3/lists/current/hardcover-fiction.json';
+const API_KEY = 'dVh2I4dlVdJYaXYaOFYtHhkGYISksZFo';
```
Comment (stpetra) [inline src/app/books/books.service.ts]: *

File: src/app/books/books.service.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+@Injectable({
+  providedIn: 'root',
+})
+export class BooksService {
+  constructor(private readonly http: HttpClient) {}
+
+  getBooks(): Observable<Book[]> {
```
Comment (stpetra) [inline src/app/books/books.service.ts]: you are missing access modifier

File: src/app/books/books.service.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    return this.http
+      .get<{ results: { books: Book[] } }>(API_URL, {
```
Comment (stpetra) [inline src/app/books/books.service.ts]: Create type for results (also no need to write _results : Type_, you can just write what is expected return type)

File: src/app/books/books.service.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        params: { 'api-key': API_KEY },
+      })
+      .pipe(
```
Comment (stpetra) [inline src/app/books/books.service.ts]: No need to write this part, httpClient method get will return Observable.

File: src/app/books/books.service.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        map(({ results }) => results.books),
+        // add delay to simulate slower network
+        delay(2000)
```
Comment (stpetra) [inline src/app/books/books.service.ts]: Why are you adding delay?

File: src/app/books/books.component.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import { Component, OnInit } from '@angular/core';
+import { BooksService } from './books.service';
+
+@Component({
+  selector: 'app-books',
+  templateUrl: './books.component.html',
+  styleUrls: ['./books.component.scss'],
+})
+export class BooksComponent implements OnInit {
+  readonly books$ = this.booksService.getBooks();
```
Comment (stpetra) [inline src/app/books/books.component.ts]: Type books$ and why is it _readonly_?

File: src/app/books/books.component.scss [hunk 8]
```diff
@@ -0,0 +1,10 @@
+.book-list {
+  margin: 2rem 1rem;
+  display: flex;
```
Comment (stpetra) [inline src/app/books/books.component.scss]: You can use FlexLayoutModule instead of 
display: flex;
flex-direction: column;
gap: 1rem;

File: src/app/books/books.component.html [hunk 9]
```diff
@@ -0,0 +1,9 @@
+<mat-toolbar color="primary">Books</mat-toolbar>
```
Comment (stpetra) [inline src/app/books/books.component.html]: I would add div or section around elements in this component

File: src/app/books/book/book.component.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import { Component, OnInit, Input } from '@angular/core';
+import { Book } from './book.model';
+
+@Component({
+  selector: 'app-book',
+  templateUrl: './book.component.html',
+  styleUrls: ['./book.component.scss'],
+})
+export class BookComponent implements OnInit {
+  @Input() book!: Book;
```
Comment (stpetra) [inline src/app/books/book/book.component.ts]: Why did you add !

File: src/app/app.module.ts [hunk 11]
```diff
@@ -1,18 +1,33 @@
...
-  declarations: [
-    AppComponent
-  ],
+  declarations: [AppComponent, BooksComponent, BookComponent],
   imports: [
     BrowserModule,
-    AppRoutingModule
+    FormsModule,
+    BrowserAnimationsModule,
+    MatToolbarModule,
```
Comment (stpetra) [inline src/app/app.module.ts]: All modules from Angular Material you can add to array and import that array here

File: src/app/app.module.ts [hunk 12]
```diff
@@ -1,18 +1,33 @@
...
+    MatCardModule,
+    MatProgressSpinnerModule,
+    CommonModule,
+    HttpClientModule,
```
Comment (stpetra) [inline src/app/app.module.ts:28]: Why are you importing HttpClientModule here?

File: src/app/books/books.component.ts (hunk 7 above)
Comment (stpetra) [reply src/app/books/books.component.ts]: Also I don't like that you have called this function here. Better solution is to call it in ngOnInit()

File: src/app/books/books.service.ts [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+@Injectable({
+  providedIn: 'root',
+})
+export class BooksService {
+  constructor(private readonly http: HttpClient) {}
+
+  getBooks(): Observable<Book[]> {
+    return this.http
+      .get<{ results: { books: Book[] } }>(API_URL, {
+        params: { 'api-key': API_KEY },
```
Comment (stpetra) [inline src/app/books/books.service.ts]: You can create
   const params: Params = {
      api-key: API_KEY,
    };
and use it here

File: src/app/app.module.ts [hunk 14]
```diff
@@ -1,18 +1,33 @@
...
+
 @NgModule({
-  declarations: [
-    AppComponent
-  ],
+  declarations: [AppComponent, BooksComponent, BookComponent],
   imports: [
     BrowserModule,
-    AppRoutingModule
+    FormsModule,
```
Comment (trebalobi) [inline src/app/app.module.ts]: Check which modules are used in the app and remove the unused ones. Also, consider importing modules on component level. Importing everything into app.module defeats the purpose of modular architecture.

File: src/app/books/books.service.ts [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+      })
+      .pipe(
+        map(({ results }) => results.books),
```
Comment (trebalobi) [inline src/app/books/books.service.ts]: You should map the observable in the component. Use service methods only for basic requests, logic should be implemented in the controller.

File: src/app/books/books.component.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+
+  constructor(private readonly booksService: BooksService) {}
+
+  ngOnInit(): void {}
```
Comment (trebalobi) [inline src/app/books/books.component.ts]: Remove readonly and initialize books$ in ngOnInit.
this.books$ = this.booksService.getBooks().pipe(map(result => result?.results?.books));

File: package.json [hunk 17]
```diff
@@ -11,13 +11,16 @@
...
+    "@angular/cdk": "^13.1.3",
     "@angular/common": "~13.1.0",
     "@angular/compiler": "~13.1.0",
     "@angular/core": "~13.1.0",
     "@angular/forms": "~13.1.0",
+    "@angular/material": "^13.1.3",
     "@angular/platform-browser": "~13.1.0",
     "@angular/platform-browser-dynamic": "~13.1.0",
     "@angular/router": "~13.1.0",
+    "hammerjs": "^2.0.8",
```
Comment (trebalobi) [inline package.json]: Why add hammerjs as a dependency?

File: src/app/books/books.component.scss [hunk 18]
```diff
@@ -0,0 +1,10 @@
...
+  flex-direction: column;
+  gap: 1rem;
```
Comment (trebalobi) [inline src/app/books/books.component.scss]: What are rem units and why use them here?

File: src/main.ts [hunk 19]
```diff
@@ -3,10 +3,12 @@ import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 
 import { AppModule } from './app/app.module';
 import { environment } from './environments/environment';
+import 'hammerjs';
```
Comment (trebalobi) [inline src/main.ts]: Why import hammerjs here?

File: src/app/books/book/book.component.html [hunk 20]
```diff
@@ -0,0 +1,11 @@
+<mat-card class="mat-elevation-z4">
+  <mat-card-title-group>
+    <mat-card-title>{{ book.title }}</mat-card-title>
```
Comment (trebalobi) [inline src/app/books/book/book.component.html]: To be absolutely sure the app wouldn't break if there is no input value for book, write: book?.title, book?.book_image, book?.description...

File: src/app/books/books.service.ts (hunk 1 above)
Comment (LjubanDragosavljevic) [reply src/app/books/books.service.ts]: I thought because of Type Inference I don't need to write it there. I will amend it.

File: src/app/books/books.service.ts (hunk 3 above)
Comment (LjubanDragosavljevic) [reply src/app/books/books.service.ts]: I read that public is default access modifier and that I don't need to write it. I will amend it.
